### ISSUE TITLE
DetailsList: Changing onRenderRow prop does not causes re-render

### ISSUE THREAD

**Issue by martlaboloa** [OfficeDev/office-ui-fabric-react#1344]
### Bug Report

- __Package version(s)__: (2.7.0)
- __Browser and OS versions__: (chrome Version 56.0.2924.87 (64-bit), Ubuntu 16.04)

#### Priorities and help requested (not applicable if asking question):

I dont think I can fix this bug myself.

Requested priority: Blocking

#### Describe the issue:

Details List component: Issue is that DetailsList decides on its own when to rerender, even after you explicitly rendered it with new props, and that kind of behaviour if fine, but i think there is bug in logic of when to rerender.

#### Actual behavior:

DetailsList ignores one of it's props, onRenderRow, meaning even if you change it, rerender does not happen, I don't exactly know which props it ignores but it defenetely ignores this one.

#### Expected behavior:

from "onRenderRow" method  i'm returning: <DetailsRow {...props} />, but wrapped in "div" tag which has some styles, so if "onRenderRow" changes and Im rendering DetailsList with new "onRenderRow" prop, it should rerender.

**Comment by dzearing** [comment]
I believe this may be the `shouldComponentUpdate` explicitly ignoring this change because of perf. Basically we're trying to execute as little redundant code as possible. If you pass in a new anything, causing a re-evaluation of the row rendering which causes a diff which done in bulk creates slow perf. So say we fix this, then others will complain about slow rendering perf.

Not saying we shouldn't fix though; we just need to figure out a way to keep perf fast. E.g. users might do this:

```tsx
<DetailsList
   onRenderRow={ (props, defaultRender) => <div>{ defaultRender(props) }</div> }
/>
```

So basically if the host ever updates, the function is recreated every time, causing re-evaluation of the list regardless of whether or not the content changes.

A temporary workaround would be to call `forceUpdate` on the list when you know it should re-evaluate. This gives you more control about when to update externally.

**Comment by martlaboloa** [comment]
yes i get it's perf. issue, **but** 'onRenderRow' has given user full controll of row rendering, and if this prop is updated, it completely changes how to render given data, sure some props must be ignored but i dont think 'onRenderRow' is one of them. Behaviour of lazy rendering, can be kept too, if u rerender only visible rows.

Specific issue that i have is that: how can i change styles of rows: colors, boldness etc..., i mean data itself is not changed, its same, but im changing 'onRenderRow' which puts styles on given rows.

**Comment by dzearing** [comment]
So the data hasn't changed, but SOMETHING changes. What is that? Is there some other metadata used outside of the items which controls this? Is it possible to observe these changes via some eventing? Because you could always have your own DetailsRow component (which can wrap the internal one) which observes that state and re-renders the individual rows as necessary. You don't need to recreate functions to pass into DetailsRow and have it re-evaluate the world, that's much less efficient.

**Comment by dzearing** [comment]
E.g. this is how react-redux would work (wrap your thing in an HOC which observes the store and re-evaluates on change.)
